### PR TITLE
scan TrashFile and PendingDeletedSlice in status subcommand

### LINKED ISSUE #3088: scan `TrashFile` and `PendingDeletedSlice` in status subcommand
**What would you like to be added**:

**Why is this needed**:

https://github.com/juicedata/juicefs/pull/3085#discussion_r1048252627

### PR DESCRIPTION
**Pull request by Hexilee**
close #3088

### CODE REVIEW

File: pkg/meta/tkv.go [hunk 1]
```diff
@@ -2304,7 +2304,49 @@ func (m *kvMeta) scanDeletedSlices(ctx Context, scan deletedSliceScan) error {
...
 
-func (m *kvMeta) scanDeletedFiles(ctx Context, scan deletedFileScan) error {
+func (m *kvMeta) scanPendingSlices(ctx Context, scan pendingSliceScan) error {
+	if scan == nil {
+		return nil
+	}
+
+	// slice refs: Kiiiiiiiissss
+	klen := 1 + 8 + 4
+	keys, err := m.scanKeys(m.fmtKey("K"))
```
Comment (davies) [inline pkg/meta/tkv.go:2314]: use scan() to fetch values, faster than `incrby(0)` for every key

File: pkg/meta/sql.go [hunk 2]
```diff
@@ -2676,7 +2676,37 @@ func (m *dbMeta) scanDeletedSlices(ctx Context, scan deletedSliceScan) error {
...
+		} else if !ok {
+			return nil
+		}
+		return tx.Where("refs <= 0").Find(&refs)
+	})
+	if err != nil {
+		return errors.Wrap(err, "scan slice refs")
+	}
+	for _, ref := range refs {
+		if ref.Refs <= 0 {
```
Comment (davies) [inline pkg/meta/sql.go:2696]: no need

File: pkg/meta/redis.go [hunk 3]
```diff
@@ -2893,7 +2893,65 @@ func (m *redisMeta) scanDeletedSlices(ctx Context, scan deletedSliceScan) error
...
+		return nil
+	}
+
+	pendingKeys := make(chan string, 1000)
+	c, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		_ = m.hscan(c, m.sliceRefs(), func(keys []string) error {
+			for i := 0; i < len(keys); i += 2 {
+				pendingKeys <- keys[i]
```
Comment (davies) [inline pkg/meta/redis.go:2907]: keys[i+1] is the value, we should check the value here

File: pkg/meta/base.go [hunk 4]
```diff
@@ -1438,6 +1442,66 @@ func (m *baseMeta) CleanupTrashBefore(ctx Context, edge time.Time, increProgress
...
+	for len(scannedEntries) != 0 {
+		poped := scannedEntries
+		scannedEntries = nil
+		for _, entry := range poped {
+			if entry.Attr.Typ == TypeFile {
+				clean, err := scan(entry.Inode, entry.Attr.Length, entry.deletedAt)
+				if err != nil {
+					return errors.Wrap(err, "scan trash files")
+				}
+				if clean {
```
Comment (SandyXSD) [inline pkg/meta/base.go:1479]: no need to clean since it's only used in `status` command, which is read only

File: pkg/meta/base.go [hunk 5]
```diff
@@ -1438,6 +1442,66 @@ func (m *baseMeta) CleanupTrashBefore(ctx Context, edge time.Time, increProgress
...
+					st = m.en.doUnlink(ctx, entry.Attr.Parent, string(entry.Name))
+					if st != 0 {
+						return errors.Wrap(st, "unlink trash file")
+					}
+				}
+				continue
+			}
+			if entry.Attr.Typ == TypeDirectory {
+				var subEntries []*Entry
+				if st = m.en.doReaddir(ctx, entry.Inode, 1, &subEntries, -1); st != 0 {
```
Comment (davies) [inline pkg/meta/base.go:1489]: The directory should be empty

File: pkg/meta/base.go [hunk 6]
```diff
@@ -81,14 +82,17 @@ type engine interface {
...
+	scanPendingFiles(Context, pendingFileScan) error
 
 	GetSession(sid uint64, detail bool) (*Session, error)
 }
 
-type deletedSliceScan func(ss []Slice, ts int64) (clean bool, err error)
-type deletedFileScan func(ino Ino, size uint64, ts int64) (clean bool, err error)
+type trashSliceScan func(ss []Slice, ts int64) (clean bool, err error)
+type pendingSliceScan func(id uint64, size uint32) (clean bool, err error)
+type trashFileScan func(inode Ino, size uint64, ts time.Time) (clean bool, err error)
```
Comment (SandyXSD) [inline pkg/meta/base.go:94]: The only thing we care here is `size`?

File: pkg/meta/base.go [hunk 7]
```diff
@@ -1438,6 +1442,66 @@ func (m *baseMeta) CleanupTrashBefore(ctx Context, edge time.Time, increProgress
...
+				continue
+			}
+			if entry.Attr.Typ == TypeDirectory {
+				var subEntries []*Entry
+				if st = m.en.doReaddir(ctx, entry.Inode, 1, &subEntries, -1); st != 0 {
+					logger.Warnf("readdir subTrash %d: %s", entry.Inode, st)
+					continue
+				}
+				for _, se := range subEntries {
+					scannedEntries = append(scannedEntries, scannedEntry{
```
Comment (SandyXSD) [inline pkg/meta/base.go:1494]: check whether the entry is file here and increase the accumulate stats if necessary, so we don't have to append files to the list.

File: pkg/meta/base.go (hunk 5 above)
Comment (Hexilee) [reply pkg/meta/base.go:1489]: Why should the directory be empty? Here we won't remove the directory, just read it and push subentries into the queue.

File: cmd/status.go [hunk 8]
```diff
@@ -64,14 +65,18 @@ type sections struct {
...
-	PendingDeletedFileSize  int64 `json:",omitempty"`
+	UsedSpace                uint64
+	AvailableSpace           uint64
+	UsedInodes               uint64
+	AvailableInodes          uint64
+	TrashSliceCount          int64 `json:",omitempty"`
+	TrashSliceSize           int64 `json:",omitempty"`
+	PendingDeletedSliceCount int64 `json:",omitempty"`
+	PendingDeletedSliceSize  int64 `json:",omitempty"`
+	TrashFileCount           int64 `json:",omitempty"`
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply cmd/status.go:76]: Reasonable.

File: pkg/meta/base.go (hunk 6 above)
Comment (Hexilee) [reply pkg/meta/base.go:94]: Currently, we only care about their's count and size.

File: pkg/meta/base.go (hunk 7 above)
Comment (Hexilee) [reply pkg/meta/base.go:1494]: Here we do a level order traversal, appending directories is necessary.

File: pkg/meta/base.go (hunk 4 above)
Comment (Hexilee) [reply pkg/meta/base.go:1479]: Ok, I prefer to keep this API, but not do actual cleaning currently.

File: pkg/meta/redis.go (hunk 3 above)
Comment (Hexilee) [reply pkg/meta/redis.go]: Done

File: cmd/status.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Hexilee) [reply cmd/status.go:72]: Done

File: cmd/status.go [hunk 9]
```diff
@@ -115,30 +120,43 @@ func status(ctx *cli.Context) error {
...
 	if ctx.Bool("more") {
 		progress := utils.NewProgress(false, false)
-		slicesSpinner := progress.AddDoubleSpinner("Trash Slices")
-		defer slicesSpinner.Done()
-		fileSpinner := progress.AddDoubleSpinner("Pending Deleted Files")
-		defer fileSpinner.Done()
-
+		trashSlicesSpinner := progress.AddDoubleSpinner("Trash Slices")
+		pendingDeletedSlicesSpinner := progress.AddDoubleSpinner("Pending Deleted Slices")
+		trashFileSpinner := progress.AddDoubleSpinner("Trash Files")
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply cmd/status.go:123]: Done

File: pkg/meta/base.go (hunk 4 above)
Comment (Hexilee) [reply pkg/meta/base.go:1479]: Done

File: pkg/meta/tkv.go (hunk 1 above)
Comment (Hexilee) [reply pkg/meta/tkv.go]: Done

File: pkg/meta/sql.go (hunk 2 above)
Comment (Hexilee) [reply pkg/meta/sql.go]: Removed

File: pkg/meta/base.go (hunk 7 above)
Comment (SandyXSD) [reply pkg/meta/base.go:1493]: append the se only when it's a directory

File: pkg/meta/base.go (hunk 5 above)
Comment (davies) [reply pkg/meta/base.go:1488]: The trash only has three levels,  trash/directory/file or empty directory. 
We should avoid the `readdir` for empty directory.

File: pkg/meta/base.go (hunk 7 above)
Comment (Hexilee) [reply pkg/meta/base.go:1493]: Refer to https://github.com/juicedata/juicefs/pull/3145#discussion_r1064307684, the directory must be empty, so we do not need level order traversal.

File: pkg/meta/base.go (hunk 5 above)
Comment (Hexilee) [reply pkg/meta/base.go]: Fixed, the new implementation ignores empty directories.

File: pkg/meta/base.go (hunk 7 above)
Comment (Hexilee) [reply pkg/meta/base.go]: Fixed as https://github.com/juicedata/juicefs/pull/3145#discussion_r1065418557

**Pull request merged**